### PR TITLE
added string-uniprot id mapping and database score threshold

### PR DESCRIPTION
**Pull request by elenikiachaki**
#39

### CODE REVIEW

File: string2pdb/README.md [hunk 1]
```diff
@@ -43,23 +44,25 @@ Ensure you have the following installed:
...
 - **Interactor**: Common name of the interacting protein
 - **Interactor_id**: STRING identifier for the interactor
+- **Interactor_UniProt_AC**: Uniprot AC of the interactor
 - **String_score**: Overall confidence score of the interaction
-- **Experimental_score**: Score from experimental evidence
-- **Database_score**: Score from database evidence
-- **Textmining_score**: Score from text-mining evidence
+- **escore**: Score from experimental evidence
+- **dscore**: Score from database evidence
+- **tscore**: Score from text-mining evidence
```
Comment (mtiberti) [inline string2pdb/README.md:68]: why the change? I think the previous names were better/more explicit

File: string2pdb/stringscore [hunk 2]
```diff
@@ -20,6 +20,70 @@ import requests
 import argparse
 import pandas as pd
 from io import StringIO
+import re
+
+# Load the aliases file:
+# I assume the file must be moved to /data/databases/ 
+aliases_df = pd.read_csv('/data/user/elenikia/9606.protein.aliases.v12.0.txt', sep='\t', header=None, names=['STRING_ID', 'Alias', 'Source'])
+
+def string2uniprot(string_id):
```
Comment (mtiberti) [inline string2pdb/stringscore:29]: `aliases_df` should be an argument here

File: string2pdb/stringscore [hunk 3]
```diff
@@ -20,6 +20,70 @@ import requests
...
+                      (aliases_df['Source'] == 'UniProt_AC')]
+
+        uniprot_acs = AC_rows['Alias'].tolist()
+
+        if not uniprot_acs:
+            print(f"Error: No Uniprot AC found for String ID {string_id}. Exiting")
+            exit(1)
+
+        # Filter UniProt ACs that match the reviewed pattern:
+        reviewed_acs = [ac for ac in uniprot_acs if reviewed_pattern.match(ac)]
```
Comment (mtiberti) [inline string2pdb/stringscore:50]: any specific reason we are filtering for these?

File: string2pdb/stringscore [hunk 4]
```diff
@@ -20,6 +20,70 @@ import requests
...
+
+        if not reviewed_acs:
+            print(f"Error: No valid reviewed UniProt AC found for STRING ID {string_id}")
+            exit(1)
+
+        if len(reviewed_acs) > 1:
+            # Here (unlike what I do when reviewed_ids >= 1), I return only the most recent one, 
+            # because the rest are either unreviewed, or older versions.
+            # Return the most recent one:
+            reviewed_acs.sort()
```
Comment (mtiberti) [inline string2pdb/stringscore:60]: are we sure that this gives us the "most recent" AC? I couldn't find if ACs are assigned chronologically

File: string2pdb/stringscore [hunk 5]
```diff
@@ -20,6 +20,70 @@ import requests
...
+    # Check if any reviewed IDs were found:
+    if not reviewed_ids:
+        print(f"Error: No valid reviewed UniProt ID found for STRING ID {string_id}")
+        exit(1)
+
+        # The script continues here if reviewed_ids >= 1.
+        # Not sure if here I should keep all of them and print them in different rows, but I think Mentha does it this way.
+        # See example of target:TP53 :For the interactor CDKN2A Mentha lists both IDs as interactors: Q8N726, P42771 in different rows
+        # In that case, it means that different proteins(not just isoforms) are encoded by same gene, so all of them
+        # should be present as different interactors.
```
Comment (mtiberti) [inline string2pdb/stringscore:83]: sounds good to me

File: string2pdb/README.md [hunk 6]
```diff
@@ -12,10 +13,10 @@ Ensure you have the following installed:
 - Python 3 (on local server: use module load python/3.10/modulefile)
 
 # ARGUMENTS
-1. `<identifier>` (required, string): HUGO Gene Name of the target protein.
+1. `<identifier>` (required, string): Uniprot Accession Code of the target protein.
 
-2. `<threshold>` (optional, float): String_score threshold. Default for MAVISp: **0.7**. 
-   To adjust the threshold, specify with `-t <threshold>`.
+2. `<threshold>` (optional, float): String_score threshold. Default for MAVISp: **String_score > 0.7** AND **Database_score** > 0.
```
Comment (mtiberti) [inline string2pdb/README.md:18]: is there any reason the database score is hardcoded in the script and not user-selectable?

**Review by mtiberti — CHANGES_REQUESTED** [review]

File: string2pdb/stringscore (hunk 4 above)
Comment (elenikiachaki) [reply string2pdb/stringscore:60]: You are absolutely right about this, there is no chronological order. I digged into the cases that do not have Ensembl_HGNC_uniprot_ids alias, and they might have 1 or more Uniprot_AC. In those cases, the id that is returned by mentha, is the one that is present in the source: UniProt_DR_neXtProt. This row is exactly one for all the String IDs. I could directly extract the Uniprot AC from there for all cases but in this way I would not account for examples that have multiple Ensembl_HGNC_uniprot_ids, for which mentha includes all of them. That's why in summary, I am parsing the Ensembl_HGNC_uniprot_ids (and retrieve as many as I find), and only in case there is none, I extract the unique UniProt_DR_neXtProt.

Comment (elenikiachaki) [reply string2pdb/stringscore:60]: Here are some examples of cases with no Ensembl_HGNC_uniprot_ids available, that are present in mentha: Q8IYH5
Q9NQM4 Q3B726 Q9ULR0 Q9H334 Q8IYA8 P15309 Q96NB1. These ids are always found in UniProt_DR_neXtProt row in the aliases file

File: string2pdb/stringscore [hunk 7]
```diff
@@ -99,20 +157,34 @@ def main():
     # Read tsv data into pandas dataframe:
     interactors_df = pd.read_csv(StringIO(interactors_response.text), sep='\t')
 
-    # Filter interactors:
-    filtered_interactors = interactors_df[interactors_df['score'] >= args.threshold]
+    # Filter interactors with both score >= threshold and dscore > 0:
+    # With database score > 0: No retrieval of unreviewed interactors
+    filtered_interactors = interactors_df[(interactors_df['score'] >= args.threshold) & (interactors_df['dscore'] > args.dthreshold)].copy()
```
Comment (mtiberti) [inline string2pdb/stringscore:162]: nitpick, any reason in particular you're `.copy()`ing?

File: string2pdb/README.md [hunk 8]
```diff
@@ -1,21 +1,28 @@
...
+   - **Experimental score**
+   - **Database score**
+   - **Text mining score**. 
+- The **String Score** threshold is adjustable, allowing the user to specify the desired confidence level for extracted interactions. 
+- The **Database Score** can be adjusted as well, to specify the desired level of association in curated databases.
+- The script also maps the STRING identifier of the retrieved interactors to their **primary UniProt ACs**, using the preprocessed 
+STRING human protein alias file (https://stringdb-downloads.org/download/protein.aliases.v12.0/9606.protein.aliases.v12.0.txt.gz).
 
-# REQUIREMENTS
+# Requirements:
```
Comment (mtiberti) [inline string2pdb/README.md:14]: could you also add what we need in terms of data? (i.e. the modified STRING database)

Comment (mtiberti) [reply string2pdb/README.md:14]: can you also upload the script for generating the modified STRING correspondence as well as instructions on what it needs and how to use it? Please remember to add a header with license information and copyright as in this script :)

File: string2pdb/README.md [hunk 9]
```diff
@@ -1,21 +1,28 @@
...
+- It searches only in the **physical** subnetwork of STRING by default, but can be adjusted to parse the entire database (functional network).
+- It provides the combined score of the confidence of each interaction (**String score**), along with 
+the three components of the physical network interaction score: 
+   - **Experimental score**
+   - **Database score**
+   - **Text mining score**. 
+- The **String Score** threshold is adjustable, allowing the user to specify the desired confidence level for extracted interactions. 
+- The **Database Score** can be adjusted as well, to specify the desired level of association in curated databases.
+- The script also maps the STRING identifier of the retrieved interactors to their **primary UniProt ACs**, using the preprocessed 
+STRING human protein alias file (https://stringdb-downloads.org/download/protein.aliases.v12.0/9606.protein.aliases.v12.0.txt.gz).
```
Comment (mtiberti) [inline string2pdb/README.md:12]: I think we should be bit more explicit on what this is (see below)

File: string2pdb/README.md [hunk 10]
```diff
@@ -1,21 +1,28 @@
...
-1. `<identifier>` (required, string): HUGO Gene Name of the target protein.
+# Arguments:
+1. `<identifier>` (required, string): Uniprot Accession Code of the target protein.
 
-2. `<threshold>` (optional, float): String_score threshold. Default for MAVISp: **0.7**. 
-   To adjust the threshold, specify with `-t <threshold>`.
+2. `<threshold>` (optional, float): String_score threshold. Default for MAVISp: **String_score > 0.7**.
+   To adjust the String_score threshold, specify with `-t <threshold>`.
 
    **Other recommended thresholds for customized searches along with the confidence of interaction:**
```
Comment (mtiberti) [inline string2pdb/README.md:27]: can we add here where these numbers come from?

**Review by mtiberti — CHANGES_REQUESTED** [review]

File: string2pdb/stringscore (hunk 7 above)
Comment (elenikiachaki) [reply string2pdb/stringscore:162]: yes, i do it to avoid getting the following warning when I try to modify the  filtered_interactors df in line 164: 
```
/data/user/elenikia/PPI2PDB/string2pdb/./stringscore:164: 
SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_interactors["Interactor_UniProt_AC"] = filtered_interactors["stringId_B"].apply(
```
